epochStart/bootstrap/epochStartDataProvider.go

```diff
@@ -0,0 +1,512 @@
+	if err != nil {
+		return nil, err
+	}
+
+	prevMetaBlock, err := esdp.getEpochStartMetaBlock(metaBlock.Epoch - 1)
+	if err != nil {
+		return nil, err
+	}
+
+	esdp.changeMessageProcessorsForMetaBlocks()
+
+	log.Info("previous meta block", "epoch", prevMetaBlock.Epoch)
+	nodesConfig, err := esdp.nodesConfigProvider.GetNodesConfigForMetaBlock(metaBlock)
+	if err != nil {
+		return nil, err
+	}
+
+	shardCoordinator, err := esdp.getShardCoordinator(metaBlock, nodesConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	shardHeaders, err := esdp.getShardHeaders(metaBlock, nodesConfig, shardCoordinator)
+	if err != nil {
+		return nil, err
+	}
+
+	epochStartData, err := esdp.getCurrentEpochStartData(shardCoordinator, metaBlock)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, mb := range epochStartData.PendingMiniBlockHeaders {
+		receivedMb, err := esdp.getMiniBlock(&mb)
+		if err != nil {
+			return nil, err
+		}
+		log.Info("received miniblock", "type", receivedMb.Type)
+	}
+
+	lastFinalizedMetaBlock, err := esdp.getMetaBlock(epochStartData.LastFinishedMetaBlock)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("received last finalized meta block", "nonce", lastFinalizedMetaBlock.Nonce)
+
+	firstPendingMetaBlock, err := esdp.getMetaBlock(epochStartData.FirstPendingMetaBlock)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("received first pending meta block", "nonce", firstPendingMetaBlock.Nonce)
+
+	trie, err := esdp.getTrieFromRootHash(epochStartData.RootHash)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ComponentsNeededForBootstrap{
+		EpochStartMetaBlock: metaBlock,
+		NodesConfig:         nodesConfig,
+		ShardHeaders:        shardHeaders,
+		ShardCoordinator:    shardCoordinator,
+		Tries:               trie,
+	}, nil
+}
+
+func (esdp *epochStartDataProvider) changeMessageProcessorsForMetaBlocks() {
+	err := esdp.messenger.UnregisterMessageProcessor(factory.MetachainBlocksTopic)
+	if err != nil {
+		log.Info("error unregistering message processor", "error", err)
+	}
+
+	err = esdp.messenger.RegisterMessageProcessor(factory.MetachainBlocksTopic, esdp.metaBlockInterceptor)
+	if err != nil {
+		log.Info("error unregistering message processor", "error", err)
+	}
+}
+
+func (esdp *epochStartDataProvider) createRequestHandler() (process.RequestHandler, error) {
+	dataPacker, err := partitioning.NewSimpleDataPacker(esdp.marshalizer)
+	if err != nil {
+		return nil, err
+	}
+
+	shardC, err := sharding.NewMultiShardCoordinator(2, core.MetachainShardId)
+	if err != nil {
+		return nil, err
+	}
+
+	storageService := &disabled.ChainStorer{
+		GetStorerCalled: func(unitType dataRetriever.UnitType) storage.Storer {
+			return disabled.NewDisabledStorer()
+		},
+	}
+
+	cacher := disabled.NewDisabledPoolsHolder()
+	triesHolder := state.NewDataTriesHolder()
+	var stateTrie data.Trie
+	// TODO: change from integrationsTests.CreateAccountsDB
+	_, stateTrie, _ = integrationTests.CreateAccountsDB(factory2.UserAccount)
```

Comment:
use directly NewDataTrie() it is easy to create

Reply:
done